File: dataload/json2neo/src/main/java/OntologyScanner.java
import com.google.gson.Gson;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonToken;

import java.io.IOException;
import java.util.*;

public class OntologyScanner {

    static Gson gson = new Gson();

    public enum NodeType {
        ONTOLOGY, CLASS, PROPERTY, INDIVIDUAL
    }
    public static class Result {
        String ontologyId;
        String ontologyUri;
        public Set<String> allOntologyProperties = new HashSet<>();
        public Set<String> allClassProperties = new HashSet<>();
        public Set<String> allPropertyProperties = new HashSet<>();
        public Set<String> allIndividualProperties = new HashSet<>();
        public Set<String> allEdgeProperties = new HashSet<>();
        public Map<String, Set<NodeType>> uriToTypes = new HashMap<>();
    }

    private static void addType(Result res, String uri, NodeType type) {
        Set<NodeType> types = res.uriToTypes.get(uri);
        if(types == null) {
            types = new TreeSet<NodeType>();
            res.uriToTypes.put(uri, types);
        }
        types.add(type);
    }

    public static Result scanOntology(JsonReader reader) throws IOException {

        Result res = new Result();

        reader.beginObject();

        while(reader.peek() != JsonToken.END_OBJECT) {

            String name = reader.nextName();

            if (name.equals("classes")) {
                reader.beginArray();
                while (reader.peek() != JsonToken.END_ARRAY) {
                    reader.beginObject(); // class
                    while(reader.peek() != JsonToken.END_OBJECT) {

                        String property = reader.nextName();

                        if(!OntologyWriter.PROPERTY_BLACKLIST.contains(property))
                            res.allClassProperties.add(property);

                        if(property.equals("iri")) {
                            addType(res, reader.nextString(), NodeType.CLASS);
                        } else {
                            Object value = gson.fromJson(reader, Object.class);
                            visitValue(property, value, res.allClassProperties, res.allEdgeProperties);
                        }
                    }
                    reader.endObject();
                }
                reader.endArray();
                continue;
            }

            if (name.equals("properties")) {
                reader.beginArray();
                while (reader.peek() != JsonToken.END_ARRAY) {
                    reader.beginObject(); // class
                    while(reader.peek() != JsonToken.END_OBJECT) {

                        String property = reader.nextName();

                        if(!OntologyWriter.PROPERTY_BLACKLIST.contains(property))
                            res.allPropertyProperties.add(property);

                        if(property.equals("iri")) {
                            addType(res, reader.nextString(), NodeType.PROPERTY);
                        } else {
                            Object value = gson.fromJson(reader, Object.class);
                            visitValue(property, value, res.allPropertyProperties, res.allEdgeProperties);
                        }
                    }
                    reader.endObject();
                }
                reader.endArray();
                continue;
            }

            if (name.equals("individuals")) {
                reader.beginArray();
                while (reader.peek() != JsonToken.END_ARRAY) {
                    reader.beginObject(); // class
                    while(reader.peek() != JsonToken.END_OBJECT) {

                        String property = reader.nextName();

                        if(!OntologyWriter.PROPERTY_BLACKLIST.contains(property))
                            res.allIndividualProperties.add(property);

                        if(property.equals("iri")) {
                            addType(res, reader.nextString(), NodeType.INDIVIDUAL);
                        } else {
                            Object value = gson.fromJson(reader, Object.class);
                            visitValue(property, value, res.allIndividualProperties, res.allEdgeProperties);
                        }
                    }
                    reader.endObject();
                }
                reader.endArray();
                continue;
            }

            if(!OntologyWriter.PROPERTY_BLACKLIST.contains(name))
                res.allOntologyProperties.add(name);

            if(name.equals("iri")) {
                res.ontologyUri = reader.nextString();
                addType(res, res.ontologyUri, NodeType.ONTOLOGY);
            } else if(name.equals("ontologyId")) {
                res.ontologyId = reader.nextString();
            } else {
                Object value = gson.fromJson(reader, Object.class);
                visitValue(name, value, res.allOntologyProperties, res.allEdgeProperties);
            }
        }

        reader.endObject();

        return res;

    }

    private static void visitValue(String predicate, Object value, Set<String> outProps, Set<String> outEdgeProps) {

        if(predicate.equals("linkedEntities")) {
            return;
        }

        if(value instanceof String) {

        } else if(value instanceof List) {

            List<Object> listValue = (List<Object>) value;

            for(Object entry : listValue)
                visitValue(predicate, entry, outProps, outEdgeProps);

        } else if(value instanceof Map) {

            // could be a typed literal, a relatedTo object, reification, a bnode,
	    // or some json junk from the ontology config

            Map<String, Object> mapValue = new TreeMap<String,Object>((Map<String, Object>) value);

	    Object type = mapValue.get("type");

	    if(type == null || ! (type instanceof List)) {
		// bnode (anon. class) or json junk
		return;
	    }

	    List<String> types = (List<String>) type;

	    if(types.contains("literal")) {

		// is this a localization?
                if(mapValue.containsKey("lang")) {

                    String lang = (String)mapValue.get("lang");
                    assert(lang != null);

                    // add a localized property like e.g. fr+http://some/predicate
                    // (english is the default and doesn't get a prefix)
                    // 
                    if(!lang.equals("en")) {
                        if(!OntologyWriter.PROPERTY_BLACKLIST.contains(predicate))
                            outProps.add(lang + "+" + predicate);
                    }
		}

	    } else if(types.contains("related")) {

                visitValue(predicate, mapValue.get("value"), outProps, outEdgeProps);

	    } else if(types.contains("reification")) {

		List<Object> axioms = (List<Object>) mapValue.get("axioms");

		for(Object axiomObj : axioms) {

		    Map<String,Object> axiom = (Map<String,Object>) axiomObj;

                    // predicates used to describe the edge itself
                    for(String edgePredicate : axiom.keySet()) {

                        if(edgePredicate.equals("type"))
                            continue;

                        if(!OntologyWriter.PROPERTY_BLACKLIST.contains(edgePredicate))
                            outEdgeProps.add(edgePredicate);
                    }
		}

                visitValue(predicate, mapValue.get("value"), outProps, outEdgeProps);

            } else if(types.contains("datatype")) {

            } else {
		throw new RuntimeException("???");
            }

        }

    }


}

File: dataload/linker/create_manifest/src/ontology_scanner.rs
use std::collections::{BTreeSet, HashSet};
use std::io::Read;

use ols_shared::streaming::read_value;
use struson::reader::{JsonReader, JsonStreamReader};

use crate::node_type::NodeType;
use crate::ontology_scan_result::OntologyScanResult;

/// Scan an ontology from a streaming JSON reader
/// This matches the behavior of Java's OntologyScanner.scanOntology
pub fn scan_ontology<R: Read>(
    reader: &mut JsonStreamReader<R>,
    ignore_properties: &HashSet<String>,
) -> OntologyScanResult {
    let mut result = OntologyScanResult::default();

    reader.begin_object().unwrap();

    while reader.has_next().unwrap() {
        let name = reader.next_name_owned().unwrap();

        match name.as_str() {
            "classes" => {
                reader.begin_array().unwrap();
                while reader.has_next().unwrap() {
                    reader.begin_object().unwrap();
                    while reader.has_next().unwrap() {
                        let property = reader.next_name_owned().unwrap();

                        if !ignore_properties.contains(&property) {
                            result.all_class_properties.insert(property.clone());
                        }

                        if property == "iri" {
                            let iri = reader.next_string().unwrap().to_string();
                            add_type(&mut result, &iri, NodeType::Class);
                        } else {
                            let value = read_value(reader);
                            visit_value(
                                &property,
                                &value,
                                &mut result.all_class_properties,
                                &mut result.all_edge_properties,
                                ignore_properties,
                            );
                        }
                    }
                    reader.end_object().unwrap();
                }
                reader.end_array().unwrap();
            }
            "properties" => {
                reader.begin_array().unwrap();
                while reader.has_next().unwrap() {
                    reader.begin_object().unwrap();
                    while reader.has_next().unwrap() {
                        let property = reader.next_name_owned().unwrap();

                        if !ignore_properties.contains(&property) {
                            result.all_property_properties.insert(property.clone());
                        }

                        if property == "iri" {
                            let iri = reader.next_string().unwrap().to_string();
                            add_type(&mut result, &iri, NodeType::Property);
                        } else {
                            let value = read_value(reader);
                            visit_value(
                                &property,
                                &value,
                                &mut result.all_property_properties,
                                &mut result.all_edge_properties,
                                ignore_properties,
                            );
                        }
                    }
                    reader.end_object().unwrap();
                }
                reader.end_array().unwrap();
            }
            "individuals" => {
                reader.begin_array().unwrap();
                while reader.has_next().unwrap() {
                    reader.begin_object().unwrap();
                    while reader.has_next().unwrap() {
                        let property = reader.next_name_owned().unwrap();

                        if !ignore_properties.contains(&property) {
                            result.all_individual_properties.insert(property.clone());
                        }

                        if property == "iri" {
                            let iri = reader.next_string().unwrap().to_string();
                            add_type(&mut result, &iri, NodeType::Individual);
                        } else {
                            let value = read_value(reader);
                            visit_value(
                                &property,
                                &value,
                                &mut result.all_individual_properties,
                                &mut result.all_edge_properties,
                                ignore_properties,
                            );
                        }
                    }
                    reader.end_object().unwrap();
                }
                reader.end_array().unwrap();
            }
            "iri" => {
                let uri = reader.next_string().unwrap().to_string();
                result.ontology_uri = uri.clone();
                add_type(&mut result, &uri, NodeType::Ontology);
                if !ignore_properties.contains("iri") {
                    result.all_ontology_properties.insert("iri".to_string());
                }
            }
            "ontologyId" => {
                result.ontology_id = reader.next_string().unwrap().to_string();
                if !ignore_properties.contains("ontologyId") {
                    result.all_ontology_properties.insert("ontologyId".to_string());
                }
            }
            _ => {
                if !ignore_properties.contains(&name) {
                    result.all_ontology_properties.insert(name.clone());
                }
                let value = read_value(reader);
                visit_value(
                    &name,
                    &value,
                    &mut result.all_ontology_properties,
                    &mut result.all_edge_properties,
                    ignore_properties,
                );
            }
        }
    }

    reader.end_object().unwrap();

    result
}

fn add_type(result: &mut OntologyScanResult, uri: &str, node_type: NodeType) {
    result
        .uri_to_types
        .entry(uri.to_string())
        .or_default()
        .insert(node_type);
}

fn visit_value(
    predicate: &str,
    value: &serde_json::Value,
    out_props: &mut BTreeSet<String>,
    out_edge_props: &mut BTreeSet<String>,
    ignore_properties: &HashSet<String>,
) {
    if predicate == "linkedEntities" {
        return;
    }

    match value {
        serde_json::Value::String(_) => {}
        serde_json::Value::Array(arr) => {
            for entry in arr {
                visit_value(predicate, entry, out_props, out_edge_props, ignore_properties);
            }
        }
        serde_json::Value::Object(map) => {
            // Could be a typed literal, a relatedTo object, reification, a bnode,
            // or some json junk from the ontology config
            let type_val = map.get("type");

            let types: Vec<String> = match type_val {
                Some(serde_json::Value::Array(arr)) => arr
                    .iter()
                    .filter_map(|v| v.as_str().map(String::from))
                    .collect(),
                _ => return, // bnode (anon. class) or json junk
            };

            if types.contains(&"literal".to_string()) {
                // Is this a localization?
                if let Some(serde_json::Value::String(lang)) = map.get("lang") {
                    // Add a localized property like e.g. fr+http://some/predicate
                    // (english is the default and doesn't get a prefix)
                    if lang != "en" && !ignore_properties.contains(predicate) {
                        out_props.insert(format!("{}+{}", lang, predicate));
                    }
                }
            } else if types.contains(&"related".to_string()) {
                if let Some(inner_value) = map.get("value") {
                    visit_value(
                        predicate,
                        inner_value,
                        out_props,
                        out_edge_props,
                        ignore_properties,
                    );
                }
            } else if types.contains(&"reification".to_string()) {
                if let Some(serde_json::Value::Array(axioms)) = map.get("axioms") {
                    for axiom_obj in axioms {
                        if let serde_json::Value::Object(axiom) = axiom_obj {
                            // Predicates used to describe the edge itself
                            for edge_predicate in axiom.keys() {
                                if edge_predicate == "type" {
                                    continue;
                                }
                                if !ignore_properties.contains(edge_predicate) {
                                    out_edge_props.insert(edge_predicate.clone());
                                }
                            }
                        }
                    }
                }
                if let Some(inner_value) = map.get("value") {
                    visit_value(
                        predicate,
                        inner_value,
                        out_props,
                        out_edge_props,
                        ignore_properties,
                    );
                }
            } else if types.contains(&"datatype".to_string()) {
                // Nothing to do for datatypes
            }
            // Other types are ignored
        }
        _ => {}
    }
}
